curvefs/src/client/s3/client_s3_cache_manager.cpp

```diff
@@ -275,69 +275,97 @@ int FileCacheManager::Read(uint64_t inodeId, uint64_t offset, uint64_t length,
-        return readOffset;
-    }
 
-    uint32_t i;
-    for (i = 0; i < totalS3Requests.size(); i++) {
-        S3ReadRequest &tmp_req = totalS3Requests[i];
-        VLOG(9) << "S3ReadRequest chunkid:" << tmp_req.chunkId
-                << ",offset:" << tmp_req.offset << ",len:" << tmp_req.len
-                << ",objectOffset:" << tmp_req.objectOffset
-                << ",readOffset:" << tmp_req.readOffset
-                << ",compaction:" << tmp_req.compaction
-                << ",fsid:" << tmp_req.fsId << ",inodeId:" << tmp_req.inodeId;
-    }
+            uint32_t i;
+            for (i = 0; i < totalS3Requests.size(); i++) {
+                S3ReadRequest& tmp_req = totalS3Requests[i];
+                VLOG(9) << "S3ReadRequest chunkid:" << tmp_req.chunkId
+                        << ",offset:" << tmp_req.offset
+                        << ",len:" << tmp_req.len
+                        << ",objectOffset:" << tmp_req.objectOffset
+                        << ",readOffset:" << tmp_req.readOffset
+                        << ",compaction:" << tmp_req.compaction
+                        << ",fsid:" << tmp_req.fsId
+                        << ",inodeId:" << tmp_req.inodeId;
+            }
 
-    std::vector<S3ReadResponse> responses;
+            std::vector<S3ReadResponse> responses;
 
-    ret = ReadFromS3(totalS3Requests, &responses, fileLen);
-    if (ret < 0) {
-        LOG(ERROR) << "read from s3 failed. ret:" << ret;
-        return ret;
-    }
+            ret = ReadFromS3(totalS3Requests, &responses, fileLen);
+            if (ret < 0) {
+                retry++;
+                if (ret != -2 || retry == maxRetry) {
+                    LOG(ERROR) << "read from s3 failed. ret:" << ret;
+                    return ret;
+                } else {
+                    // ret -2 refs s3obj not exist
+                    // clear inodecache && get again
+                    LOG(INFO) << "inode cache maybe steal, try to get latest";
+                    inodeManager->ClearInodeCache(inodeId);
+                    auto r = inodeManager->GetInode(inodeId, inodeWrapper);
+                    if (r != CURVEFS_ERROR::OK) {
+                        LOG(WARNING) << "get inode fail, ret:" << ret;
+                        return -1;
+                    }
+                    continue;
+                }
+            }
 
-    auto repIter = responses.begin();
-    for (; repIter != responses.end(); repIter++) {
-        VLOG(6) << "readOffset:" << repIter->GetReadOffset()
-                << ",bufLen:" << repIter->GetBufLen();
-        memcpy(dataBuf + repIter->GetReadOffset(), repIter->GetDataBuf(),
-               repIter->GetBufLen());
+            auto repIter = responses.begin();
+            for (; repIter != responses.end(); repIter++) {
+                VLOG(6) << "readOffset:" << repIter->GetReadOffset()
+                        << ",bufLen:" << repIter->GetBufLen();
+                memcpy(dataBuf + repIter->GetReadOffset(),
```

Comment:
Should we move this memcpy outside the while loop?

Reply:
if success, loop will break, it's ok.

Reply:
fixed